File: internal/collect/logs.go
package collect

// Logs collector.
//
//	/log/print    the backlog, once, at connect
//	/log/listen   a PUSH STREAM: RouterOS sends each entry as it is written
//
// THE ONLY STREAMING COLLECTOR IN THE PORT, and the one place where a stream is
// clearly right. Everything else here polls, because a poll costs one channel
// for a moment and a stream holds one open — but a log has no "current state" to
// read: polling /log/print would mean re-reading the whole buffer every tick and
// keeping a seen-set to work out what was new. The listen channel makes both
// unnecessary, which is exactly what the Node original says it is for.
//
// The history is a RING BUFFER, not a growing slice. A busy router writes
// continuously, and the page shows a window; keeping everything would be a leak
// with a log level attached to it.

import (
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"mikrodash/internal/routeros"
)

// Declared as a Cmd so the proplist drift gate can compare it against what
// logs.js asks for. /log/listen carries no proplist — it pushes whole rows.
var (
	logPrintCmd = routeros.Cmd{Path: "/log/print", Args: []string{"=.proplist=time,topics,message"}}
	// SET B: A STREAM, and the only one identified by its path rather than by an
	// argument. See acquisition.go.
	logListenCmd = routeros.Cmd{Path: "/log/listen"}
)

// Streamer is the half of a router connection that keeps a channel open.
//
// Separate from Reader because ONE collector needs it and the replay harness
// cannot provide it: a fixture records answers to reads, and a recorded stream
// is replayed as successive reads instead. A Reader that does not implement this
// simply gets no live tail, which is the honest degradation — the backlog still
// loads and the page still renders.
type Streamer interface {
	Stream(routeros.Cmd, func(routeros.Reply)) (func(), error)
}

// LogEntry is one line as the page renders it.
type LogEntry struct {
	// TS is when THIS PROCESS saw the entry, not when the router wrote it —
	// `Time` is the router's own stamp, in the router's timezone, with no
	// offset. Both travel because the page sorts on one and shows the other.
	TS       int64  `json:"ts"`
	Time     string `json:"time"`
	Topics   string `json:"topics"`
	Message  string `json:"message"`
	Severity string `json:"severity"`
}

// logHistorySize is the ring buffer's depth, from the same environment variable
// the Node collector reads so a deployment that raised it keeps its setting.
func logHistorySize() int {
	if v := os.Getenv("LOG_HISTORY_SIZE"); v != "" {
		// parseInt semantics, matching the original: a leading number wins and
		// anything unparseable falls back to the default rather than to zero.
		if n, err := strconv.Atoi(strings.TrimSpace(v)); err == nil && n > 0 {
			return n
		}
	}
	return 500
}

// classifyLog maps RouterOS's topic list to the four severities the page styles.
//
// SUBSTRING MATCHING ON THE WHOLE LIST, in this order, because a row carries
// several topics at once — "system,error,critical" is one string, and the first
// branch that matches wins. Reproduced exactly: reordering these would reclassify
// rows, and the page colours them.
func classifyLog(topicsRaw string) string {
	t := strings.ToLower(topicsRaw)
	switch {
	case strings.Contains(t, "critical") || strings.Contains(t, "error"):
		return "error"
	case strings.Contains(t, "warning"):
		return "warning"
	case strings.Contains(t, "debug"):
		return "debug"
	}
	return "info"
}

// Logs is the collector.
type Logs struct {
	ros  Reader
	emit Emit
	size int

	mu      sync.Mutex
	history []LogEntry
	stop    func()
}

// The two rooms this collector serves. The page and the dashboard card show the
// same lines, and a viewer can hold both.
// logRooms is kept as the name this file has always used; the VALUE now comes
// from the one declaration in rooms.go, so the guard and the emit agree.
var logRooms = logsRooms.Join()

func NewLogs(ros Reader, emit Emit) *Logs {
	return &Logs{ros: ros, emit: emit, size: logHistorySize()}
}

// FoldLog folds one log row into the history ring and returns the entry it made
// and the next state.
//
// ── PHASE 4.1: THE SECOND SEQUENCE DERIVATION, SAME SHAPE AS THE FIRST ─────
//
//	table     func(prior, []Reply) (payload, prior)   map over the current state
//	sequence  func(prior,   Reply) (payload, prior)   fold one element in
//
// A log line is the clearest case of the second: each row is an EVENT, it
// matters once, and there is no "current value of the log" to map over. That is
// also why this menu can never back a rolling cache entry -- see
// `roscache.unrollable`, where it is refused for exactly this reason.
//
// PURE, AND THE COPY IS NOT DEFENSIVE. `append` into a slice with spare capacity
// writes THROUGH to the caller's backing array, and these rings always have
// spare capacity once they have been sliced down. `FoldPing`'s own test passed a
// mutation for a whole round because it was written with slice literals, whose
// capacity equals their length; this avoids the same trap by construction.
func FoldLog(prior []LogEntry, size int, row routeros.Reply, now int64) (LogEntry, []LogEntry) {
	topics := row["topics"]
	e := LogEntry{
		TS: now, Time: row["time"], Topics: topics,
		Message: row["message"], Severity: classifyLog(topics),
	}
	return e, appendCapped(prior, e, size)
}

// appendCapped is the ring rule, in one place.
//
// THE LAST `size`, not the first: a ring that has overflowed must yield its most
// RECENT lines. Trimming from the front is what makes it a ring rather than a
// buffer that stops accepting, and it is the same rule `LoadInitial` applies to
// the backlog -- `/log/print` answers oldest first, so a router with more history
// than the ring holds must give up its oldest.
//
// It COPIES rather than appending in place. A slice with spare capacity -- which
// every one of these has once it has been trimmed -- would otherwise be written
// through, and the caller's view of its own history would change under it.
func appendCapped(prior []LogEntry, e LogEntry, size int) []LogEntry {
	next := append(append(make([]LogEntry, 0, len(prior)+1), prior...), e)
	if size > 0 && len(next) > size {
		next = next[len(next)-size:]
	}
	return next
}

// push appends one entry, dropping the oldest when the ring is full.
//
// The collector's half: hand the carried ring to the fold and keep what comes
// back. Everything that can be got wrong lives in FoldLog.
func (l *Logs) push(e LogEntry) {
	l.history = appendCapped(l.history, e, l.size)
}

func entryOf(row routeros.Reply, now int64) LogEntry {
	e, _ := FoldLog(nil, 0, row, now)
	return e
}

// LoadInitial reads the backlog and emits it whole.
//
// The LAST `size` rows, not the first: /log/print answers oldest first, and a
// router with more history than the ring holds must yield its most recent lines,
// not its oldest. A row with no message is dropped rather than rendered blank.
func (l *Logs) LoadInitial() {
	if !l.ros.Connected() {
		return
	}
	rows, err := l.ros.Do(logPrintCmd)
	if err != nil {
		return
	}
	if len(rows) > l.size {
		rows = rows[len(rows)-l.size:]
	}
	now := time.Now().UnixMilli()
	l.mu.Lock()
	for _, row := range rows {
		if row["message"] == "" {
			continue
		}
		l.push(entryOf(row, now))
	}
	out := l.snapshot()
	l.mu.Unlock()
	EvLogsHistory.Emit(l.emit, logRooms, out)
}

// Listen opens the push channel. A Reader that cannot stream gets the backlog
// and nothing further, which is what the replay harness sees.
func (l *Logs) Listen() {
	s, ok := l.ros.(Streamer)
	if !ok || !l.ros.Connected() {
		return
	}
	l.mu.Lock()
	if l.stop != nil {
		l.mu.Unlock()
		return
	}
	l.mu.Unlock()

	stop, err := s.Stream(logListenCmd, func(row routeros.Reply) {
		if row["message"] == "" {
			return
		}
		e := entryOf(row, time.Now().UnixMilli())
		l.mu.Lock()
		l.push(e)
		l.mu.Unlock()
		// ONE ENTRY PER FRAME, not the whole history. A busy router writes
		// several lines a second and the page appends; re-sending the buffer
		// each time would be the same data over and over.
		EvLogsNew.Emit(l.emit, logRooms, e)
	})
	if err != nil {
		return
	}
	l.mu.Lock()
	l.stop = stop
	l.mu.Unlock()
}

func (l *Logs) Start() {
	l.LoadInitial()
	l.Listen()
}

// Reconnected drops the buffer and reloads. The router that came back may have
// rebooted, in which case its log starts again and the lines held here describe
// a different uptime.
//
// EXPRESSED THROUGH Resume, so the two paths cannot drift: coming back from a
// reconnect and coming back from a suspend need the same three things in the
// same order, and they used to be written out twice.
func (l *Logs) Reconnected() {
	l.Stop()
	l.Resume()
}

func (l *Logs) Stop() {
	l.mu.Lock()
	stop := l.stop
	l.stop = nil
	l.mu.Unlock()
	if stop != nil {
		stop()
	}
}

// Suspend gives up the push channel.
//
// ── IT WAS A NO-OP FOR THE WHOLE LIFE OF THE PORT ──────────────────────────
//
// `logs` was the one page-fed collector nothing could stop. It held
// `/log/listen` open from connect to teardown on every router, whether or not
// anybody had ever opened the Logs page -- and because the page switchboard only
// ran from a frame the browser sent, an empty pair of methods here looked like a
// deliberate design rather than a gap. The comment that justified it said the
// channel was the point; what it did not say is that the channel is exactly the
// resource this project conserves.
//
// ── THE TRADE, STATED, BECAUSE IT IS NOT FREE ──────────────────────────────
//
// Suspending costs a `/log/print` on the next resume: the ring develops a gap
// while nothing is listening, so the backlog has to come from the router again.
// That is one command per visit, which is what every other page-gated collector
// already pays on focus -- and CLAUDE.md is explicit that fewer channels is the
// win and command count is not the bottleneck. A router that nobody watches the
// logs of now holds no channel for them at all.
func (l *Logs) Suspend() { l.Stop() }

// Resume re-opens the channel, and RELOADS FIRST.
//
// The ring is dropped rather than appended to. `push` does not deduplicate, so
// reading `/log/print` on top of a ring that still holds the same lines would
// show every one of them twice; and a ring with a gap in the middle is worse
// than an empty one, because the page renders it as continuous.
//
// IDEMPOTENT, and that is required rather than tidy: `applyDemand` calls
// `ResumeCollector` for every wanted collector on every focus and blur, so a
// viewer flipping between pages reaches this several times a minute. Reloading
// each time would be a `/log/print` per navigation.
func (l *Logs) Resume() {
	l.mu.Lock()
	running := l.stop != nil
	if running {
		l.mu.Unlock()
		return
	}
	l.history = nil
	l.mu.Unlock()
	l.Start()
}

// snapshot copies the ring under the caller's lock.
func (l *Logs) snapshot() []LogEntry {
	out := make([]LogEntry, len(l.history))
	copy(out, l.history)
	return out
}

// Last is the history, for the page replay and for the differential gate.
func (l *Logs) Last() []LogEntry {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.history == nil {
		return nil
	}
	return l.snapshot()
}

File: web/src/pages/logs.ts
// The Logs page — a port of the `── Logs` region in public/app.js.
//
// The only page fed by a PUSH STREAM. `logs:history` arrives once with the
// backlog; `logs:new` arrives one line at a time as the router writes it. The
// two paths are deliberately different and both are reproduced:
//
//   history  rebuilds the buffer and re-renders the whole view
//   new      appends ONE line to the DOM and trims the head
//
// Re-rendering everything on each arrival would be simpler and wrong: a busy
// router writes several lines a second, and rebuilding a two-thousand-line view
// that often loses the scroll position, the text selection, and any hope of
// reading it.
//
// ── EVERY LINE IS PRE-RENDERED ──────────────────────────────────────────────
//
// The buffer holds `{html, severity, text}`, not the entry. The HTML is built
// once on arrival and the lowercased search text with it, so filtering is a
// string test over a prepared array rather than an escape-and-format pass per
// keystroke. That is the live design and it is why the search stays responsive
// at two thousand lines.

import { esc, el, debounce } from '../dom';
import type { Socket } from '../socket';
import type { LogEntry } from '../gen/payloads';

interface Buffered { html: string; severity: string; text: string }

const MAX_LOG_LINES = 2000;

// A topic gets a colour so the eye can group without reading. Substring
// matching on the whole list, in this order — a line carries several topics and
// the first branch that matches wins.
function topicClass(t: string): string {
  const s = String(t).toLowerCase();
  if (s.includes('firewall') || s.includes('forward')) return 'log-firewall';
  if (s.includes('dhcp')) return 'log-dhcp';
  if (s.includes('wireless') || s.includes('wifi') || s.includes('wlan')) return 'log-wireless';
  if (s.includes('system')) return 'log-system';
  return 'log-topic';
}

function sevClass(s: string): string {
  return s === 'error' ? 'log-error'
    : s === 'warning' ? 'log-warning'
      : s === 'debug' ? 'log-debug' : 'log-info';
}

function buildLogHtml(l: LogEntry): string {
  return '<div class="log-line"><span class="log-time">' + esc(l.time) + '</span> ' +
    '<span class="' + topicClass(l.topics) + '">[' + esc(l.topics) + ']</span> ' +
    '<span class="' + sevClass(l.severity) + '">' + esc(l.message) + '</span></div>';
}

export function initLogsPage(socket: Socket, isVisible: (page: string) => boolean): void {
  const logsEl = el('logs');
  const logSearch = el<HTMLInputElement>('logSearch');
  const logSeverity = el<HTMLSelectElement>('logSeverity');
  const toggleScroll = el('toggleScroll');
  const clearLogs = el('clearLogs');
  if (!logsEl) return;

  const logCountEls: Record<string, HTMLElement | null> = {
    error: el('logCountError'), warning: el('logCountWarning'),
    info: el('logCountInfo'), debug: el('logCountDebug'),
  };

  let logBuffer: Buffered[] = [];
  let autoScroll = true;
  let logFilter = '';
  let logLevel = '';

  function updateLogCounts(): void {
    const counts: Record<string, number> = { error: 0, warning: 0, info: 0, debug: 0 };
    logBuffer.forEach((e) => { if (counts[e.severity] !== undefined) counts[e.severity]!++; });
    Object.keys(counts).forEach((sev) => {
      const e = logCountEls[sev];
      if (!e) return;
      const n = counts[sev]!;
      // "1 error" but "2 errors", and `info`/`debug` never pluralise — they are
      // the level's name, not a count of things with a plural.
      e.textContent = n + ' ' + (sev === 'error' && n !== 1 ? 'errors'
        : sev === 'warning' && n !== 1 ? 'warnings' : sev);
    });
  }

  function flushLogs(): void {
    const f = logBuffer.filter((e) => {
      if (logLevel && e.severity !== logLevel) return false;
      if (logFilter && e.text.indexOf(logFilter) === -1) return false;
      return true;
    });
    logsEl!.innerHTML = f.map((e) => e.html).join('');
    if (autoScroll) logsEl!.scrollTop = logsEl!.scrollHeight;
    updateLogCounts();
  }

  function bufferedOf(line: LogEntry): Buffered {
    return {
      html: buildLogHtml(line),
      severity: line.severity,
      // Lowercased ONCE, on arrival: the search runs over this, not over the
      // rendered HTML, so a filter cannot match a class name or an escape.
      text: (line.time + ' [' + line.topics + '] ' + line.message).toLowerCase(),
    };
  }

  // Initialise the badge labels to "0 …" immediately, so the header is not blank
  // before the first payload.
  updateLogCounts();

  socket.on('logs:history', (data) => {
    logBuffer = data.map(bufferedOf);
    if (logBuffer.length > MAX_LOG_LINES) {
      logBuffer.splice(0, logBuffer.length - MAX_LOG_LINES);
    }
    flushLogs();
  });

  socket.on('logs:new', (line) => {
    const entry = bufferedOf(line);
    logBuffer.push(entry);
    if (logBuffer.length > MAX_LOG_LINES) logBuffer.shift();
    // The COUNTS move even when the line is filtered out — a hidden error is
    // still an error, and the badge is how you find out it happened.
    updateLogCounts();
    if (logLevel && entry.severity !== logLevel) return;
    if (logFilter && entry.text.indexOf(logFilter) === -1) return;
    logsEl!.insertAdjacentHTML('beforeend', entry.html);
    while (logsEl!.children.length > MAX_LOG_LINES) {
      logsEl!.removeChild(logsEl!.firstElementChild!);
    }
    if (autoScroll) logsEl!.scrollTop = logsEl!.scrollHeight;
  });

  logSearch?.addEventListener('input', debounce(() => {
    logFilter = (logSearch.value || '').trim().toLowerCase();
    flushLogs();
  }, 200));

  logSeverity?.addEventListener('change', () => {
    logLevel = logSeverity.value;
    Object.keys(logCountEls).forEach((s) => {
      logCountEls[s]?.classList.toggle('active', s === logLevel);
    });
    flushLogs();
  });

  toggleScroll?.addEventListener('click', () => {
    autoScroll = !autoScroll;
    toggleScroll.textContent = autoScroll ? 'Pause' : 'Resume';
  });

  // Clear empties the VIEW and the buffer, not the router's log. The next
  // history replay refills it, which is the intended escape hatch: it is a way
  // to stop reading what has already been read.
  clearLogs?.addEventListener('click', () => {
    logBuffer = [];
    logsEl!.innerHTML = '';
    updateLogCounts();
  });

  // A badge doubles as a filter toggle: clicking the active one clears it.
  Object.keys(logCountEls).forEach((sev) => {
    const e = logCountEls[sev];
    if (!e) return;
    e.addEventListener('click', () => {
      if (logLevel === sev) {
        logLevel = '';
        if (logSeverity) logSeverity.value = '';
      } else {
        logLevel = sev;
        if (logSeverity) logSeverity.value = sev;
      }
      Object.keys(logCountEls).forEach((s) => {
        logCountEls[s]?.classList.toggle('active', s === logLevel);
      });
      flushLogs();
    });
  });

  void isVisible;
}
